### ISSUE TITLE
Serializing Parameters

### ISSUE THREAD

**Issue by deevashwer** [shaih/HElib#207]
Is there an option in HElib to serialize the FHEcontext class after building the parameters?
I am working with parameters that take a lot of time to get initialized. Is there another way I can get around this problem?

**Comment by fionser** [comment]
To write out all the data associated with a context, do the following:
  \code
    writeContextBase(str, context);
    str << context;
  \endcode

**Comment by deevashwer** [comment]
Thanks @fionser for your answer. 
I did this to serialize my parameters, but the problem is that it is taking almost as much time as the original computation took.
Initializing a FHEcontext object, even after giving gens and ords, takes a lot of time.
After that I add the serialized information about the modulo chain, which doesn't take much time.
The problem is that initializing the FHEcontext object takes a lot of time, even after serialization.

**Comment by deevashwer** [comment]
That's exactly what I did, as I mentioned above. I initialized it with gens and ords as well, yet it is taking too much time. Almost as much as the original computation took to generate gens and ords.

**Comment by deevashwer** [comment]
So, I benchmarked the performance for a parameter setting that took 356 seconds to initialize the FHEcontext object the first time (without providing ords and gens). 
Code: `FHEcontext context(m, p, r));`
Then, I serialized the parameters.
This time it took 335 seconds, even after reading the Context base, and providing the same to the FHEcontext object. It took almost the same amount of time. 
Code: `FHEcontext context(m, p, r, gens, ords));
contextFile >> context;
`
This totally defeats the purpose of serialization.

**Comment by fionser** [comment]
@deevashwer how about providing your `m` and `p` and `L` ?

**Comment by deevashwer** [comment]
My parameters are: p=(7593119, 7593127, 7593133), r=2, m=65536, l=8, c=2 
(I have instantiated 3 instances of FHEcontext)

The part that is taking a lot of time is the initialization of FHEcontext object. So, I assume that only m, p and r are relevant.

**Comment by shaih** [comment]
The serialization/deserialization of context in HElib is mostly an afterthought. The use case that I had in mind was that the context is initialized "at the beginning of time" and stays fixed thereafter. Serialization/deserializations is written to ensure that you can send a context across the wire and get the same thing on the other end, it was not written to be efficient. Most structures are re-computed rather than saved and read back. Specifically the FHEcontext constructor re-initalizes from scratch the underlying PAlgebra, PAlgebraMod, and EncryptedArray, each of which consists of computing many interpolation coefficients etc.

If anyone has the time to put in re-writing it, let me know and I'll be happy to give you some guidance on what needs to be changed and how.

**Comment by deevashwer** [comment]
I am currently dealing with a use case where I am using HElib, without leveraging its slots. I'm dealing with a polynomial plaintext directly, so I used the encryption and decryption methods of publicKey and secretKey directly, to save encoding and decoding time.
Is there a way I can initialize parameters quickly, since most of the computation done by FHEcontext is not being used by me?

**Comment by fionser** [comment]
@shaih In my understanding, the `PAlgebra.cpp` do many table generation work, so we can serialize these tables (i.e., `genMaskTable` and `genCrtTable`) ?

**Comment by shaih** [comment]
Yes, both `PAlgebra` and `PAlgebraMod` have many tables, which can be serialized.

In `PAlgebra` the tables are `T`, `Tidx`, `zmsIdx`, `zmsRep`, `native`, and also `PhimX`. (Probably the most expensive to initialize is the table `T`, then `zmsIndx` and `PhimX`, once these are set it should not take very long to intialize the others, so maybe it is enough to record only these tables.)

In `PAlgebraModDerived` the tables are `factors`,  `crtCoeffs`, `maskTable`, `crtTable`, `crtTree`, I'm not sure which of them is the most important one to initialize, we can put some timers in the constructor to see what takes the longest to compute.

The `EncryptedArray` object also has some tables, but these take a lot faster to initialize.

**Comment by fionser** [comment]
@shaih, I am working on this. If I can get a reasonable form without too many modification, I will send a PR.

**Comment by shaih** [comment]
@fionser, I just merged an implementation of "binary I/O", which saves on bandwidth when you serlialize/deserialize stuff.

This is mostly orthogonal to the issue of re-computing all these tables, but if you find a good solution for the ascii I/O then we should also copy it to the binary case.

**Comment by fionser** [comment]
@shaih It seems I need to modify too much codes of FHEcontext, PAlgebra to serialize the tables, e.g., creating a new dummy constructor of PAlgebra, and to restore the tables from file. 

I might need to rethink about how to do the that.

By the way, the current binary IO dump data as little endian format, so would it be fine to work on big-endian machine ?
